project-api-tests/src/test/java/org/ballerina/projectapi/BalToolTest.java

```diff
@@ -0,0 +1,870 @@
+        }
+    }
+
+    @Test(description = "List all tools when there are no tools", groups = {"list"})
+    public void testListToolsWhenNoToolsInstalled() throws IOException, InterruptedException {
+        // remove the tool entirely
+        executeToolCommand(DISTRIBUTION_FILE_NAME, tempWorkspaceDirectory,
+                new ArrayList<>(Arrays.asList("remove", toolId)),
+                toolEnvironments.get(ToolSubCommand.LIST).envVariables);
+
+        // list all tools
+        Process cmdExec = executeToolCommand(DISTRIBUTION_FILE_NAME, tempWorkspaceDirectory,
+                new ArrayList<>(List.of("list")), toolEnvironments.get(ToolSubCommand.LIST).envVariables);
+        String cmdErrors = getString(cmdExec.getErrorStream());
+        if (!cmdErrors.isEmpty()) {
+            Assert.fail(OUTPUT_CONTAIN_ERRORS + cmdErrors);
+        }
+
+        // Validate the command output
+        String cmdOutput = getString(cmdExec.getInputStream());
+        String expectedOutput = readExpectedCmdOutsAsString("tool-list-with-no-tools.txt");
+        if (!cmdOutput.contains(expectedOutput)) {
+            Assert.fail(OUTPUT_NOT_CONTAINS_EXP_MSG + expectedOutput + "\nactual output:" + cmdOutput);
+        }
+    }
+
+    @Test(description = "List all tools when there are multiple versions of tools", groups = {"list"})
+    public void testListToolsWithMultipleToolVersions() throws IOException, InterruptedException {
+        // Pull the latest tool version
+        executeToolCommand(DISTRIBUTION_FILE_NAME, tempWorkspaceDirectory,
+                new ArrayList<>(Arrays.asList("pull", toolId)), toolEnvironments.get(ToolSubCommand.LIST).envVariables);
+
+        // Pull tool version with specific tool version
+        String specificToolVersion = toolId + ":" + specificVersion;
+        executeToolCommand(DISTRIBUTION_FILE_NAME, tempWorkspaceDirectory,
+                new ArrayList<>(Arrays.asList("pull", specificToolVersion)),
+                toolEnvironments.get(ToolSubCommand.LIST).envVariables);
+
+        // list all tools
+        Process cmdExec = executeToolCommand(DISTRIBUTION_FILE_NAME, tempWorkspaceDirectory,
+                new ArrayList<>(List.of("list")), toolEnvironments.get(ToolSubCommand.LIST).envVariables);
+        String cmdErrors = getString(cmdExec.getErrorStream());
+        if (!cmdErrors.isEmpty()) {
+            Assert.fail(OUTPUT_CONTAIN_ERRORS + cmdErrors);
+        }
+
+        // Validate the command output
+        String cmdOutput = getString(cmdExec.getInputStream());
+        String expectedOutput = readExpectedCmdOutsAsString("tool-list-with-multiple-tool-versions.txt");
+        if (!cmdOutput.contains(expectedOutput)) {
+            Assert.fail(OUTPUT_NOT_CONTAINS_EXP_MSG + expectedOutput + "\nactual output:" + cmdOutput);
+        }
+    }
+
+    @Test(description = "Search a tool with tool id", groups = {"list"})
+    public void testSearchAToolWithId() throws IOException, InterruptedException {
```

Comment:
Let's add a test case to verify the output when an invalid pattern is given to the search command. e.g `disttest-0.1.0`, `disttest/0.1.0`, `disttest 0.1.0`

Reply:
Since we are allowing any text other than tool id as well for `bal tool search` similar to `bal search`. Therefore we are passing the outputs that are not tool-ids to the central as well. Central checks if the text matches any orgs, names, keywords etc as well.